mne/io/base.py

```diff
@@ -2672,232 +2626,242 @@ def __exit__(self, exc_type, exc_value, traceback):
+    reset_range: bool = field(init=False)
+    data_type: int = field(init=False)
 
-    # previous file name and id
-    if part_idx > 0 and prev_fname is not None:
-        start_block(fid, FIFF.FIFFB_REF)
-        write_int(fid, FIFF.FIFF_REF_ROLE, FIFF.FIFFV_ROLE_PREV_FILE)
-        write_string(fid, FIFF.FIFF_REF_FILE_NAME, prev_fname)
-        if info["meas_id"] is not None:
-            write_id(fid, FIFF.FIFF_REF_FILE_ID, info["meas_id"])
-        write_int(fid, FIFF.FIFF_REF_FILE_NUM, part_idx - 1)
-        end_block(fid, FIFF.FIFFB_REF)
-
-    pos_prev = fid.tell()
-    if pos_prev > split_size:
-        raise ValueError(
-            'file is larger than "split_size" after writing '
-            "measurement information, you must use a larger "
-            "value for split size: %s plus enough bytes for "
-            "the chosen buffer_size" % pos_prev
+    def __post_init__(self):
+        type_dict = dict(
+            short=FIFF.FIFFT_DAU_PACK16,
+            int=FIFF.FIFFT_INT,
+            single=FIFF.FIFFT_FLOAT,
+            double=FIFF.FIFFT_DOUBLE,
         )
+        _check_option("fmt", self.fmt, type_dict.keys())
+        reset_dict = dict(short=False, int=False, single=True, double=True)
+        object.__setattr__(self, "reset_range", reset_dict[self.fmt])
+        object.__setattr__(self, "data_type", type_dict[self.fmt])
+
+
+class _RawFidWriter:
+    def __init__(self, raw, info, picks, projector, start, stop, cfg):
+        self.raw = raw
+        self.picks = _picks_to_idx(info, picks, "all", ())
+        self.info = pick_info(info, self.picks)
+        self.projector = projector
+        self.start, self.stop = start, stop
+        self.cfg = cfg
+
+    def write(self, fid, part_idx, prev_fname, next_fname):
+        start_block(fid, FIFF.FIFFB_MEAS)
+        self._start_writing_raw(fid)
+        cals = self._get_calibrations()
+        is_next_split = self._write_raw_fid(fid, cals, part_idx, prev_fname, next_fname)
+        end_block(fid, FIFF.FIFFB_MEAS)
+        return is_next_split
+
+    def _write_raw_fid(self, fid, cals, part_idx, prev_fname, next_fname):
+        self._start_raw_block(fid)
+
+        self._check_start_stop_within_bounds()
+        first_samp = self.raw.first_samp + self.start
+        if first_samp != 0:
+            write_int(fid, FIFF.FIFF_FIRST_SAMPLE, first_samp)
+
+        # previous file name and id
+        if part_idx > 0 and prev_fname is not None:
+            self._write_neighbour_fname(fid, prev_fname, part_idx - 1, "prev")
+
+        pos_prev = fid.tell()
+        if pos_prev > self.cfg.split_size:
+            raise ValueError(
+                'file is larger than "split_size" after writing '
+                "measurement information, you must use a larger "
+                "value for split size: %s plus enough bytes for "
+                "the chosen buffer_size" % pos_prev
+            )
 
-    # Check to see if this has acquisition skips and, if so, if we can
```

Comment:
This block and the one below it are about 100 lines of what I assume is mostly (only?) just an indentation change. To make `diff` on this PR and `blame` going forward more manageable, could you see if putting this in a private `def _write_raw_fid_func(...):` or similar would work to keep the diff minimal? Then your `_RawFidWriter._write_raw_fid` could just call that function with whatever arguments it needs.

Perhaps you could also keep `_start_writing_raw` as it was in a similar way?

I agree that improving the flow control and removing recursion would be great. At the same time there is a price to pay for changing indentation in terms of `blame` working properly (or not) and the diff growing bigger. So if we could reduce the `diff` to see just the class-based simplifications better (leaving the low-level writing stuff more alone) then I think it could be a good compromise...

Reply:
Ok, I see your point. Let me do it like this then.